R/pool.R

```diff
@@ -19,164 +18,238 @@ NULL
+          private$changeObjectStatus(id, object, state, NULL)
+          stop("Object could not be returned back to the pool. ",
+               "It was destroyed instead. Error message: ",
+               conditionMessage(e))
         })
-        attr(object, "reapTaskHandle") <- taskHandle
+
+        ## set up a task to destroy the object after `idleTimeout`
+        ## millis, if we're over the minimum number of objects
+        taskHandle <- scheduleTask(
+          self$idleTimeout, function() {
+            if (self$counters$free + self$counters$taken > self$minSize) {
+              private$changeObjectStatus(id, object, "free", NULL)
+            }
+          }
+        )
+        attr(object, "pool_metadata")$destroyHandle <- taskHandle
+
+        private$changeObjectStatus(id, object, "taken", "free")
+
+        ## set up recurring validation every `validationInterval` millis
+        ## so we can catch if an idle connection gets broken somehow
+        ## (but only if we have a proper scheduler)
+        if (!(is.null(getOption("pool.scheduler", NULL)))) {
+          pool_metadata <- attr(object, "pool_metadata", exact = TRUE)
+          pool_metadata$validateHandle <-
+            scheduleTaskRecurring(self$validationInterval, function() {
+              object <- private$checkValid(object)
+              ## if we got here, the object was successfully
+              ## activated and validated; now needs to be passivated
+              onPassivate(object)
+            })
+        }
       })
     },
-    ## cleaning up and closing the pool
-    close = function() {                      #### WHAT TO DO IF THIS THROWS AN ERROR???
-      self$valid <- FALSE
-      freeEnv <- private$freeObjects
-      ## destroy all objects
-      eapply(freeEnv, private$destroyObject)
-      ## empty the objects' environments
-      rm(list = ls(freeEnv), envir = freeEnv)
-      ## TODO: deal with leaked connections, since there's no longer a takenEnv to check
+
+    ## cleaning up and closing the pool -- after the pool
+    ## is closed, objects that were previously checked out
+    ## can still be returned to the pool (which will
+    ## immediately destroy them). Objects can no longer be
+    ## checked out from the pool.
+    close = function() {
+      naiveScheduler$protect({
+        if (!self$valid) stop("The pool was already closed.")
+
+        self$valid <- FALSE
+        freeEnv <- private$freeObjects
+        freeObs <- ls(freeEnv)
+
+        ## destroy all free objects
+        for (id in freeObs) {
+          obj <- freeEnv[[id]]
+          private$cancelScheduledTask(obj, "destroyHandle")
+          private$changeObjectStatus(id, obj, "free", NULL)
+        }
+
+        # check if there are taken objects
+        if (self$counters$taken > 0) {
+          warning("You still have checked out objects. Return ",
+                  "them to the pool so they can de destroyed. ",
```

Comment:
Sorry, that's shorthand from sed and vim. https://robots.thoughtbot.com/sed-102-replace-in-place

This shorthand used to be pretty common on technical forums, maybe not so much anymore.